src/sentry/search/snuba/base.py

```diff
@@ -0,0 +1,732 @@
+                id__in=[gid for gid, _ in snuba_groups]
+            ).values_list("id", flat=True)
+
+            group_to_score = dict(snuba_groups)
+            for group_id in filtered_group_ids:
+                if group_id in result_group_ids:
+                    # because we're doing multiple Snuba queries, which
+                    # happen outside of a transaction, there is a small possibility
+                    # of groups moving around in the sort scoring underneath us,
+                    # so we at least want to protect against duplicates
+                    continue
+
+                group_score = group_to_score[group_id]
+                result_group_ids.add(group_id)
+                result_groups.append((group_id, group_score))
+
+        # TODO do we actually have to rebuild this SequencePaginator every time
+        # or can we just make it after we've broken out of the loop?
+        paginator_results = SequencePaginator(
+            [(score, id) for (id, score) in result_groups], reverse=True, **paginator_options
+        ).get_result(limit, cursor, known_hits=hits)
+
+        # break the query loop for one of three reasons:
+        # * we started with Postgres candidates and so only do one Snuba query max
+        # * the paginator is returning enough results to satisfy the query (>= the limit)
+        # * there are no more groups in Snuba to post-filter
+        if candidate_ids or len(paginator_results.results) >= limit or not more_results:
+            break
+
+    # HACK: We're using the SequencePaginator to mask the complexities of going
+    # back and forth between two databases. This causes a problem with pagination
+    # because we're 'lying' to the SequencePaginator (it thinks it has the entire
+    # result set in memory when it does not). For this reason we need to make some
+    # best guesses as to whether the `prev` and `next` cursors have more results.
+
+    if len(paginator_results.results) == limit and more_results:
+        # Because we are going back and forth between DBs there is a small
+        # chance that we will hand the SequencePaginator exactly `limit`
+        # items. In this case the paginator will assume there are no more
+        # results, so we need to override the `next` cursor's results.
+        paginator_results.next.has_results = True
+
+    if cursor is not None and (not cursor.is_prev or len(paginator_results.results) > 0):
+        # If the user passed a cursor, and it isn't already a 0 result `is_prev`
+        # cursor, then it's worth allowing them to go back a page to check for
+        # more results.
+        paginator_results.prev.has_results = True
+
+    metrics.timing("snuba.search.num_chunks", num_chunks)
+
+    groups = Group.objects.in_bulk(paginator_results.results)
+    paginator_results.results = [groups[k] for k in paginator_results.results if k in groups]
+
+    return paginator_results
+
+
+class SnubaSearchBackendBase(SearchBackend):
```

Comment:
It seems to me there will be three implementations here:
1) old system that calls postgres and then Snuba based on the result
2) New system
2.1) Fully Snuba backend: this will skip postgres and call snuba on joined data
2.2) New system that behaves like the old one for some specific queries.

Are you planning to have three backends: 1, 2.2, 2.2 each one extending SearchBackend 
or do you plan to have a common parent class between 1 and 2 and 2 being a parent class for 2.1 and 2.2 ?

Reply:
There are two implementations.
1. Old, unchanged (albeit refactored), system.
2. The updated one which tries to use snuba more wherever possible - migrating more and more as new data is available in snuba. Currently, this means now being able to query certain group parameters using the groups dataset (like status, last seen, active at, etc) in snuba rather than postgres.

They share a common base class, but the base class right now is essentially 1 (the class of 1 just extends the base class and doesn't add anything on top of it).
2 also extends the base class, but imports only the variables it needs, while overriding any functions (and declaring any variables) it needs to be different.

Reply:
ok